### PR TITLE
CSHARP-2353: BsonDocument.DeepClone() fails on documents with duplicate element names

### PR DESCRIPTION
**Pull request by BorisDog**
Also addresses [CSHARP-2354](https://jira.mongodb.org/browse/CSHARP-2354)

### CODE REVIEW

**Review by rstam — CHANGES_REQUESTED** [review]
Visual Studio warned that the line endings in BsonDocument.cs are inconsistent.

Other than that everything looks OK.

File: src/MongoDB.Bson/ObjectModel/BsonDocument.cs [hunk 1]
```diff
@@ -68,6 +68,22 @@ public BsonDocument(BsonElement element)
...
+        /// <summary>
+        /// Initializes a new instance of the BsonDocument by coping elements from another BsonDocument.
+        /// </summary>
+        /// <param name="document">The document whose elements will be copied</param>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors")]
+        public BsonDocument(BsonDocument document)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException("document");
```
Comment (damieng) [inline src/MongoDB.Bson/ObjectModel/BsonDocument.cs:80]: `nameof(document)` instead of `"document"` ensures any refactoring stays in tact.

**Comment by BorisDog** [comment]
> Visual Studio warned that the line endings in BsonDocument.cs are inconsistent.
> 
> Other than that everything looks OK.

Thanks, line endings should be consistent now.

File: src/MongoDB.Bson/ObjectModel/BsonDocument.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Bson/ObjectModel/BsonDocument.cs]: Thanks! Done.

**Review by rstam — APPROVED** [review]
LGTM

**Pull request merged**